zips/zip-0231.md

```diff
@@ -0,0 +1,332 @@
+require zero-knowledge proofs, and for the Orchard protocol these proofs are too
+large to fit into a 512-byte memo field.
+
+Something about light client truncation (simplifies the security argument).
+
+In the light client protocol, a recipient need not download full transaction
+information if the compact transaction information indicates that they do not
+receive any memo in the transaction.
+
+
+# Requirements
+
+- Recipients can receive memo data that is greater than 512 bytes in length.
+- Multiple recipients, across any of the shielded pools, can be given the
+  capability to view the same memo data.
+- The exact number and exact lengths of distinct decryptable memos should not
+  be revealed, even to the transaction recipients, although an upper bound on
+  the total length of memo data that the observer does not have the capability
+  to view will be leaked.
+- A recipient can determine whether or not they have been given the capability
+  to view any memo solely by decrypting the note ciphertext.
+- Memo chunks within a transaction can be individually pruned without preventing
+  the transaction from being verified.
+  - As a consequence, a recipient with the capability to view a memo can prune
+    all of that memo's chunks.
+
+
+# Non-requirements
+
+- Recipients do not need to be able to receive multiple memos per note.
+
+
+# Specification
+
+## Memo bundle
+
+A memo bundle consists of a sequence of 256-byte memo chunks, each individually
+encrypted. These memo chunks represent zero or more encrypted memos.
+
+Each transaction may contain a single memo bundle, and a memo bundle may contain
+at most 40 memo chunks. This limits the total amount of memo data that can be
+conveyed within a single transaction to 10 KiB.
+
+Memo bundles are encoded in transactions in a prunable manner: each memo chunk
+can be replaced by its representative digest.
+
+## Memo encryption
+
+During transaction construction, each memo is assigned a 32-byte `memo_key`.
+These keys SHOULD be generated randomly, and MUST NOT be used to encrypt more
+than one memo within a single transaction.
+
+In shielded bundles, the 512-byte memo field is replaced by `[flag_byte] || memo_key`.
+
+TODO: Define `flag_byte` format. Have one bit for "there is no memo".
+
+The transaction builder generates a 32-byte nonce from a CSPRNG. A new nonce MUST be
+generated for each memo bundle.
+
+The symmetric encryption key for a memo is derived from its `memo_key` as follows:
+
+    encryption_key = PRF^{expand}_{memo_key}([0xE0] || nonce)
+
+Each memo is padded to a multiple of 256 bytes with zeroes, and split into
+256-byte chunks. Each memo chunk is encrypted with ChaCha20Poly1305 as follows:
+
+    IETF_AEAD_CHACHA20_POLY1305(encryption_key, nonce = I2BEOSP_{88}(counter) || [final_chunk], memo_chunk)
+
+This is a variant of the STREAM construction [#stream]_.
+- `counter` is a big-endian chunk counter starting at zero and incrementing by
```

Comment:
I intentionally use a big-endian counter and a flag byte at the end, to match every other STREAM variant (that I'm aware of). See https://c2sp.org/age#payload and links therein.